sklearn/inspection/_h_statistic.py

```diff
@@ -0,0 +1,325 @@
+"""Friedman and Popescu's H-Statistic"""
+
+# Authors: The scikit-learn developers
+# SPDX-License-Identifier: BSD-3-Clause
+
+import itertools
+
+import numpy as np
+from scipy import sparse
+
+from ..base import is_classifier, is_regressor
+from ..utils import Bunch, check_array
+from ..utils._indexing import _get_column_indices, _safe_assign, _safe_indexing
+from ..utils._param_validation import (
+    HasMethods,
+    Integral,
+    Interval,
+    Real,
+    validate_params,
+)
+from ..utils.random import sample_without_replacement
+from ..utils.validation import _check_sample_weight, check_is_fitted
+
+
+def _calculate_pd_brute_fast(
+    pred_fun, X, feature_indices, grid, sample_weight=None, reduce_binary=False
+):
+    """Fast version of _calculate_partial_dependence_brute()
+
+    Returns np.array of size (n_grid, output_dim).
+    """
+
+    # X is stacked n_grid times, and grid columns are replaced by replicated grid
+    n = X.shape[0]
+    n_grid = grid.shape[0]
+
+    X_stacked = _safe_indexing(X, np.tile(np.arange(n), n_grid), axis=0)
```

Comment:
So I assume that the speed up observed between this function `_calculate_pd_brute_fast` and `_partial_dependence_brute` is only related to stacking all samples in a single matrix and call `.predict_proba` a single time.

So basically, we have a trade-off speed/memory consumption. Here, we might blow up the memory with large dataset and if we decide to not `subsample`.

Reply:
To ease maintenance, I'm really leaning towards using the `_partial_dependence_brute` implementation.

However, the pattern here shows that we can have a good speed-up by concatenating data but we probably need to think about a chunking strategy to not blow up memory.